includes/class-wc-stripe-api.php

```diff
@@ -341,4 +345,32 @@ public static function detach_payment_method_from_customer( string $customer_id,
 			'payment_methods/' . $payment_method_id . '/detach'
 		);
 	}
+
+	/**
+	 * Check if we should detaches a payment method from the given customer. If the site is a staging site in live
+	 * mode, we should not detach the payment method from the customer to avoid detaching it from the production site.
+	 *
+	 * @param string $customer_id        The ID of the customer that contains the payment method that should be detached.
+	 * @param string $payment_method_id  The ID of the payment method that should be detached.
+	 *
+	 * @return bool  True if we should detach the payment method, false otherwise.
+	 */
+	public static function should_detach_payment_method_from_customer() {
+		$options   = get_option( 'woocommerce_stripe_settings' );
+		$test_mode = isset( $options['testmode'] ) && 'yes' === $options['testmode'];
+
+		// If we are in test mode, we can always detach the payment method.
+		if ( $test_mode ) {
+			return true;
+		}
+
+		// Return false for the delete user request from the admin dashboard when the site is a staging site.
+		// This is to avoid detaching the payment method from the production site.
+		// Requests coming from the customer account page i.e delete payment method, are not affected by this and returns true.
+		if ( 'staging' === wp_get_environment_type() && is_admin() ) {
```

Comment:
This does not protect from detaching a payment method from the local dev copy of the production site though. Alternatively, we can use the following condition-

```
if ( is_admin() ) {
			if ( 'production' === wp_get_environment_type() ) {
				return true;
			} else {
				return false;
			}
		}
```

This however will keep the payment method attached to the customer when the delete user request is from a local dev environment which is in Stripe live mode, irrespective of the site being a copy of a production site or a completely local dev site.

Reply:
I've done some digging around and it seems like some plugins use `wp_get_environment_type()` somewhat consistently.

WC core uses the Jetpack::Status class which has a `is_staging_site()` function. That function has fallback mechanisms which use a list of known url regex patterns essentially to check if the url includes some form of `.staging.com` for example. 

However when it uses `wp_get_environment_type()` it considers "staging" to be any type that is not `'production'` or `'local'`. Which by process of elimination means `staging = 'development' or 'staging'`.

WooPayments also uses `wp_get_environment_type()` as well in [`WCPay\Core\Mode`](https://github.com/Automattic/woocommerce-payments/blob/6.1.0/includes/core/class-mode.php#L86). In that class it also considers dev sites to be `development` or `staging` environments. 

I don't think we need to copy them per se, I guess it just depends on what we want to do in each environment. 

IMO I'd think we'd only want to detach payment methods on production sites. Having said that, I don't know if I can distinguish between `development`, `local` or `staging` in this case though. It sounds like you might have an idea on that though with what you said here:

> This however will keep the payment method attached to the customer when the delete user request is from a local dev environment which is in Stripe live mode, irrespective of the site being a copy of a production site or a completely local dev site.
 
To me all three `development`, `local` and `staging` indicate a duplicate copy of the site for testing purposes and a site we wouldn't want to send live destructive requests from.

Reply:
Thanks @james-allan for the references. 

> From your comment above, it sounds like you're leaning towards local dev sites legitimately detaching payment methods from customers. Is that right?

I initially made the comment to seek the reviewer's opinion as I was unsure which one was the best solution. 
After going through the references you mentioned, I am leaning towards detaching the method from the production site only as in other places also staging and development sites are considered the same.

> To me all three development, local and staging indicate a duplicate copy of the site for testing purposes and a site we wouldn't want to send live destructive requests from.

This is also a good point.

I have made the changes in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2672/commits/5e5aae8d89b0496393ddab268e76794fe05469db accordingly.